### PR TITLE
feat(acp): emit plan sessionUpdates from ferment lifecycle

### LINKED ISSUE #1036: Plans in Kimchi ACP (kimchi-native-chat #05)
## What to build

Kimchi emits ACP `Plan` sessionUpdates with structured steps derived from its ferment lifecycle. AO's chat UI renders plan progress for Kimchi sessions — the same plan display it already shows for Claude Code.

## Status

In-review (PR #1034). Cross-reference: PR #1030 (Tautvydas) implements the same ticket with a simpler architecture — subscribe to todo store directly, no `FERMENT_EVENTS` dependency.

## Key distinction: Plans ≠ Todos ≠ Ferments

- **Plan** — the output of plan mode. Structured markdown with Goal/Constraints/Chunks. Lifecycle: drafted → completed (PLAN_COMPLETE) → approved → executed.
- **Todo** — general task-tracking the model uses during any work. Informal, ad-hoc, not specific to planning.
- **Ferment** — a separate concept. Plans can be converted to ferments, but a ferment is its own thing with phases, steps, verification gates, grading.

ACP `Plan` maps to Kimchi's ferment lifecycle — that's the only path with deterministic structured status tracking.

## Signal path

The ACP server subscribes to `FERMENT_EVENTS.PHASE_STARTED`. When a phase starts: read the active ferment via `getActive()`, map phases/steps to `PlanEntry[]` (all `pending`), set `activePlan = { planId: ferment.id, entries }`, emit initial `plan` sessionUpdate.

After the initial plan is emitted, subscribe to `subscribeTodoStore()` for status changes. The ferment todo-sync bridge maintains structured todos mapped from ferment step lifecycle (STEP_STARTED → in_progress, STEP_COMPLETED → completed, STEP_FAILED → blocked).

## Gaps from PR #1030

PR #1030 (Tautvydas) implements the same ticket independently. PR #1034 (ours) has the stronger architecture (encapsulated `AcpPlanTracker`, deduplication, error resilience, execute-path gating, structured plan content, V2-ready `planId`), but PR #1030 identified four gaps that our PR does not address. These should be implemented as follow-ups.

### Gap 1: Session resume (`loadSession`)

**Problem:** `loadSession` (session resume) doesn't emit an initial plan snapshot. A user who resumes a session mid-ferment sees no plan in the IDE until the next todo store change.

**Fix:** In `loadSession`, after restoring the todo store, emit one initial plan snapshot from restored ferment-scoped todos. Gate on having an active ferment (not just any todos) — only ferment-scoped todos should produce a plan.

**Acceptance:** `loadSession` with restored ferment-scoped todos emits exactly one `plan` sessionUpdate with correct statuses. `loadSession` with no restored todos or only global-scope todos emits nothing.

### Gap 2: `activeForm` as plan entry content

**Problem:** `in_progress` entries show the static `content` (e.g. `"write tests"`) instead of the present-continuous `activeForm` (e.g. `"writing tests"`).

**Fix:** In `todoToPlanEntry`, use `todo.activeForm ?? todo.content` as the entry content when `status === "in_progress"`.

**Acceptance:** An `in_progress` todo with `activeForm` set shows the `activeForm` as content. An `in_progress` todo without `activeForm` falls back to `content`.

### Gap 3: `blocked` UX — content marker vs silent status mapping

**Problem:** `blocked → in_progress` silently shows a blocked item as "in progress" in the IDE, which is misleading. The user can't tell why the item is stuck.

**Fix:** Map `blocked → "pending"` (v1 has no `failed` status) and prepend a `[blocked]` marker to the content, with the note appended if present. Example: `[blocked] deploy — waiting on ops`. This makes the blocked state visible in the checklist without relying on a status v1 doesn't have.

**Acceptance:** A `blocked` todo maps to `status: "pending"` with content `"[blocked] {content}"`. A `blocked` todo with a `note` appends `" — {note}"`. Unit test verifies both cases.

**Note:** The acceptance criteria below reflects this updated mapping.

### Gap 4: E2E tests

**Problem:** PR #1034 has unit tests but no end-to-end tests exercising the real ACP binary.

**Fix:** Add `tests/e2e/acp/plan-updates.test.ts` with at least two scenarios:
1. Create/update/clear lifecycle — create todos via scripted model, assert plan snapshots across create → update → clear.
2. Session isolation — two sessions, only the owning session sees plan notifications.

Use the existing `startAcpFixture` + scripted model responses pattern from other ACP e2e tests.

**Acceptance:** `pnpm run test:e2e:acp` passes with the new plan-updates test. Full ACP e2e suite remains green.

## Acceptance criteria

- [ ] When `FERMENT_EVENTS.PHASE_STARTED` fires, the ACP server emits an initial `plan` sessionUpdate with all entries set to `pending`.
- [ ] Subsequent todo store changes produce updated `plan` sessionUpdates with correct statuses (`pending` → `in_progress` → `completed`).
- [ ] `TodoStatus "blocked"` maps to ACP `"pending"` with a `[blocked]` content marker and note appended (see Gap 3).
- [ ] Priority defaults to `"medium"` for all entries.
- [ ] Only ferment-scoped todos (not global-scope) are included in plan entries.
- [ ] Execute path without ferment: no `plan` sessionUpdate is emitted.
- [ ] Unit tests verify: initial plan emission with correct entries; status updates from todo store changes; mapping of `blocked` → `pending` with marker; exclusion of global-scope todos.
- [ ] **Gap 1:** `loadSession` emits an initial plan snapshot from restored ferment-scoped todos.
- [ ] **Gap 2:** `in_progress` entries use `activeForm` as content when available.
- [ ] **Gap 3:** `blocked` todos show `[blocked]` marker + note instead of silent `in_progress` mapping.
- [ ] **Gap 4:** E2E tests in `tests/e2e/acp/plan-updates.test.ts` cover create/update/clear lifecycle + session isolation.

### PR DESCRIPTION
**Pull request by cast-vytautas**
## Linked issue

Closes #1036

## What does this PR do?

Implements ticket #05 — Plans in Kimchi ACP, including the four follow-up review gaps from the linked issue.

**Core (ticket scope):**
- The ACP server emits `plan` sessionUpdates with structured steps derived from the ferment lifecycle (`AcpPlanTracker`, `src/modes/acp/plans.ts`).
- Subscribes to `FERMENT_EVENTS.PHASE_STARTED` for initial plan emission and `subscribeTodoStore()` for status updates.
- Maps ferment-scoped todos to `PlanEntry[]`; session-correlated so plans are emitted only for the owning session.
- Hardened against callback failures and double-start (idempotent start/stop, dedupe key reset).

**Follow-up gaps:**
1. **Session-resume snapshots** — `emitRestoredSnapshot()` emits the current plan when a session is loaded with an active ferment (`loadSessionFresh`), gated on a non-empty ferment phase scope.
2. **`activeForm` content** — in-progress entries prefer `todo.activeForm` for display content.
3. **Blocked semantics** — blocked todos surface as `pending` with `[blocked] {content} — {note}` marker (ACP has no `blocked` status).
4. **E2E coverage** — `tests/e2e/acp/plan-updates.test.ts` drives create/update/clear lifecycle snapshots and session isolation through a real `kimchi --mode acp` binary against a seeded ferment (snapshot + full genesis events log so the event-store fold works from cold start).

Also: `tests/e2e/acp/support/acp-fixture.ts` now dumps the full client/server state artifact (`*.fail.acp-e2e.log`) on **any** test failure via `onTestFailed`, making CI failures self-diagnosing.

## Test plan

- `pnpm vitest run src/modes/acp/plans.test.ts` — 22/22 unit tests
- `pnpm run test:e2e:acp` — 15/15 including the 2 new plan lifecycle scenarios
- `pnpm run typecheck`, `pnpm run lint` — clean
- CI: ACP E2E, CI, TUI E2E — all green

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md) and agree to the CLA
- [x] This PR links to an open issue above
- [x] Tests pass locally (`pnpm run test`)
- [x] Lint passes (`pnpm run check`)
- [x] Documentation updated if behavior changed

### CODE REVIEW

File: src/modes/acp/plans.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		this.activePlan = plan
+		this.options.onActivePlanChanged?.(plan)
+	}
+
+	private onPhaseStarted(raw: unknown): void {
+		const payload = raw as FermentPhaseStartedPayload
+		const ferment = this.getActiveFerment()
+		// Guard mirrors the todo-sync bridge: ignore events for a ferment that
+		// isn't the currently active one.
+		if (!ferment || ferment.id !== payload.fermentId) return
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:323]: Fixed in commit 92e284a0: lastEmittedKey is now set only after a successful send().

File: src/modes/acp/plans.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+	constructor(private readonly options: AcpPlanTrackerOptions) {
+		this.getActiveFerment = options.getActiveFerment ?? getActive
+	}
+
+	start(): void {
+		// The todo-sync bridge subscribes to the bus at session_start (inside
+		// bindExtensions), so by the time the server starts the tracker its
+		// PHASE_STARTED handler runs *before* this one for the same event —
+		// the bridge's phase-scope write is already in the store and fires no
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:239]: Fixed in commit 92e284a0: onActivePlanChanged and send() are wrapped in try/catch with stderr logging.

File: src/modes/acp/plans.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+			entries.push(todoToPlanEntry(todo))
+		}
+		for (const step of phase.steps) {
+			const stepKey = getTodoScopeKey({ kind: "ferment-step", phaseId: phase.id, stepId: step.id })
+			for (const todo of state.byScope[stepKey]?.todos ?? []) {
+				entries.push(todoToPlanEntry(todo))
+			}
+		}
+	}
+	return entries
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:203]: Fixed in commit 92e284a0: start()/stop() are now idempotent via a started flag.

File: src/modes/acp/plans.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	start(): void {
+		// The todo-sync bridge subscribes to the bus at session_start (inside
+		// bindExtensions), so by the time the server starts the tracker its
+		// PHASE_STARTED handler runs *before* this one for the same event —
+		// the bridge's phase-scope write is already in the store and fires no
+		// rebuild from us (activePlan is still undefined), leaving our initial
+		// all-pending emission as the first plan the client sees.
+		if (this.options.events) {
+			this.unsubscribeEvents = this.options.events.on(FERMENT_EVENTS.PHASE_STARTED, (raw) => this.onPhaseStarted(raw))
+		}
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:247]: Deferred: the event bus is internal and typed; runtime validation would be over-defensive for an internal event.

File: src/modes/acp/plans.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			this.onTodoStoreChanged(emitterSessionId),
+		)
+	}
+
+	stop(): void {
+		this.unsubscribeEvents?.()
+		this.unsubscribeEvents = undefined
+		this.unsubscribeTodos?.()
+		this.unsubscribeTodos = undefined
+		this.setActivePlan(undefined)
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:265]: Deferred: tracking as a future enhancement when ferment teardown lifecycle is finalized.

File: src/modes/acp/plans.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+ * wired the events bus) and stop it when the session is disposed.
+ */
+export class AcpPlanTracker {
+	private activePlan: ActivePlan | undefined
+	private unsubscribeEvents: (() => void) | undefined
+	private unsubscribeTodos: (() => void) | undefined
+	private lastEmittedKey = ""
+	private readonly getActiveFerment: () => Ferment | undefined
+
+	constructor(private readonly options: AcpPlanTrackerOptions) {
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.ts:227]: Fixed in commit 15b7dc5a: lastEmittedKey is reset to empty string in stop().

File: src/modes/acp/plans.test.ts [hunk 7]
```diff
@@ -0,0 +1,394 @@
+// Unit tests for the ACP plan tracker (ticket #05): ferment lifecycle →
```
[the remark this answers is not in the document]
Comment (cast-vytautas) [reply src/modes/acp/plans.test.ts:1]: Deferred: callback failure tests were added in commit 92e284a0; remaining edge cases tracked for future work.

**Review by tautvydasLiekis — APPROVED** [review]
